### PR TITLE
feature/session-activity

### PR DESCRIPTION
**Pull request by builtbywill**
…now timeout, and keep alive is sent on-demand with user activity (mouse, keyboard, touch, etc.)

### CODE REVIEW

**Comment by gavin-brown** [comment]
What would happen if someone were to watch a video for 35 minutes (or some other value above the session end threshold)? Would they expire?

**Comment by builtbywill** [comment]
@gavin-brown In theory that could be possible, if they watched a video and did not move the mouse or anything. Possibly could hook into the <video> or <audio> element events, but I don't think those bubble up. Also, that wouldn't work for flash or embedded video.

I did expose the `onActivity` method, so if this was implemented with other functionality like video, I suppose you could trigger that manually based on the video controls.

File: lib/caliper-service.js [hunk 1]
```diff
@@ -436,27 +404,100 @@ function CaliperService(options) {
...
+				return audio.currentTime > 0 && !audio.paused && !audio.ended && audio.readyState > 2;
+			});
+		return isVideoPlaying || isAudioPlaying;
+	}
+
+	function startSessionEndTimeout() {
+		clearSessionEndTimeout();
+		logger('startSessionEndTimeout');
+		sessionEndTimeout = setTimeout(function() {
+			if (isMediaPlaying()) {
```
Comment (builtbywill) [inline lib/caliper-service.js]: @gavin-brown I was able to find a way to check for html5 video and audio elements playing.

Comment (jkdufair) [inline lib/caliper-service.js]: Now they can put on auto-play and show the prof they were watching for HOURS

File: lib/caliper-service.js [hunk 2]
```diff
@@ -180,37 +178,48 @@ function CaliperService(options) {
...
-			startSession(extensionsCopy);
+		// resume session
+		if (moment().diff(moment(lastActivityDate)) <= _options.sessionEndThreshold) {
+			restartSessionEndTimeout();
+			return;
 		}
+
+		// close previous session
+		// persist person, session.extensions
+		var extensionsCopy = session.extensions;
```
Comment (jkdufair) [inline lib/caliper-service.js]: At the risk of being pedantic, these aren't copies. They are more like saved references. Might be worth renaming.

File: lib/caliper-service.js [hunk 3]
```diff
@@ -436,27 +404,100 @@ function CaliperService(options) {
...
+		window.addEventListener('touchstart', onActivity);
+		window.addEventListener('touchmove', onActivity);
+		window.addEventListener('touchend', onActivity);
+		window.addEventListener('touchcancel', onActivity);
+		window.addEventListener('scroll', onActivity);
+		window.addEventListener('wheel', onActivity);
+		window.addEventListener('beforeunload', onActivity);
+	}
+
+	function isMediaPlaying() {
```
Comment (jkdufair) [inline lib/caliper-service.js]: 👍

**Comment by jkdufair** [comment]
Looks great to me.

File: lib/caliper-service.js (hunk 2 above)
Comment (builtbywill) [inline lib/caliper-service.js]: renamed 👍

**Pull request merged**